sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py

```diff
@@ -52,76 +122,187 @@ def _get_component_hash(cls, component: Component) -> str:
 
-        # directly resolve node if the resolution involves no remote call
-        # so that no node will be registered when create_or_update a subgraph
-        if isinstance(component, str):
-            node._component = self._resolve_component(component)  # pylint: disable=protected-access
-            return
-        if component.id is not None:
-            node._component = component.id  # pylint: disable=protected-access
-            return
+        # TODO: calculate hash without resolving additional includes (copy code to temp folder)
+        # note that it's still thread-safe with current implementation, as only read operations are
+        # done on the original code folder
+        with component._resolve_local_code() as code:  # pylint: disable=protected-access
+            if code is None or code._is_remote:  # pylint: disable=protected-access
+                return in_memory_hash
+
+            if hasattr(code, "_upload_hash"):
+                content_hash = code._upload_hash  # pylint: disable=protected-access
+            else:
+                path = code.path if os.path.isabs(code.path) else os.path.join(code.base_path, code.path)
+                if os.path.exists(path):
+                    content_hash = get_object_hash(path)
+                else:
+                    # this will be gated by schema validation, so it shouldn't happen except for mock tests
+                    return in_memory_hash
+
+            object_hash = hashlib.sha256()
+            object_hash.update(in_memory_hash.encode("utf-8"))
+
+            object_hash.update(content_hash.encode("utf-8"))
+            return _CODE_INVOLVED_PREFIX + object_hash.hexdigest()
+
+    @staticmethod
+    def get_on_disk_cache_base_dir() -> Path:
+        """Get the base path for on disk cache."""
+        from azure.ai.ml._version import VERSION
+        return Path(tempfile.gettempdir()).joinpath(".azureml", "azure-ai-ml", VERSION, "cache", "components")
+
+    def _get_on_disk_cache_path(self, on_disk_hash: str) -> Path:
+        """Get the on disk cache path for a component."""
+        return self.get_on_disk_cache_base_dir().joinpath(self._workspace_hash, on_disk_hash)
 
-        # 1 possible optimization is to save an id(component) -> component hash map here to
-        # avoid duplicate hash calculation. The risk is that we haven't implicitly forbidden component
-        # modification after it's been used in a node, and we can't guarantee that the hash is still
-        # valid after modification.
-        component_hash = self._get_component_hash(component)
-        if component_hash not in self._component_cache:
-            self._hash_of_components_to_resolve.append(component_hash)
-            self._component_cache[component_hash] = component
+    def _load_from_on_disk_cache(self, on_disk_hash: str) -> Optional[str]:
+        """Load component arm id from on disk cache."""
+        on_disk_cache_path = self._get_on_disk_cache_path(on_disk_hash)
```

Comment:
mark: should disk cache expire? Or should we do some specific error handling to let user disable on-disk cache?